OpenRA.Mods.Common/AI/BaseBuilder.cs

```diff
@@ -37,14 +42,44 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 			playerPower = pm;
 			playerResources = pr;
 			this.category = category;
+			cachedBases = 0;
+			waterAvailable = false;
+			checkedWater = false;
+			failCount = 0;
+			failRetryTicks = ai.Info.StructureProductionResumeDelay;
 		}
 
 		public void Tick()
 		{
+			// If failed to place something N consecutive times, wait M ticks until resuming building production
+			if (failCount >= ai.Info.MaximumFailedPlacementAttempts && failRetryTicks > 0)
+				--failRetryTicks;
+			else if (failCount >= ai.Info.MaximumFailedPlacementAttempts && failRetryTicks <= 0)
+				failCount = 0;
+
 			// Only update once per second or so
 			if (--waitTicks > 0)
 				return;
 
+			if (!checkedWater)
+			{
+				waterAvailable = ai.EnoughWaterToBuildNaval();
+				checkedWater = true;
+			}
+
+			if (checkedWater && !waterAvailable)
```

Comment:
Since you're doing this check in Tick anyway, you might as well have it check against the currently available build area, and reset it each time a building is placed or lost.